main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -114,35 +114,48 @@ object Kinder {
    */
   private def visitStruct(struct0: ResolvedAst.Declaration.Struct, taenv: Map[Symbol.TypeAliasSym, KindedAst.TypeAlias], root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Struct, KindError] = struct0 match {
     case ResolvedAst.Declaration.Struct(doc, ann, mod, sym, tparams0, fields0, loc) =>
-      // if not annotated tparams are assumed to be Star except the last one which is Eff
-      val tparams1 = tparams0 match {
+      // In the case in which the user doesn't supply any type params,
+      // the parser will have already notified the user of this error
+      // The recovery step here is to simply add a single type param that is never used
+      val tparams1 = if (tparams0.tparams.isEmpty) {
+        val regionTparam = ResolvedAst.TypeParam.Unkinded(Name.Ident("", loc), Symbol.freshUnkindedTypeVarSym(Ast.VarText.Absent, true, loc), loc)
+        ResolvedAst.TypeParams.Unkinded(List(regionTparam))
+      } else {
+        tparams0
+      }
+      // if not annotated, tparams are assumed to be Star except the last one which is Eff
+      val tparams2 = tparams1 match {
         case ResolvedAst.TypeParams.Kinded(t) => ResolvedAst.TypeParams.Kinded(t)
         case ResolvedAst.TypeParams.Unkinded(t) =>
           val tparams = t.init.map(tparam => ResolvedAst.TypeParam.Kinded(tparam.name, tparam.sym, Kind.Star, tparam.loc)) :+ ResolvedAst.TypeParam.Kinded(t.last.name, t.last.sym, Kind.Eff, t.last.loc)
           ResolvedAst.TypeParams.Kinded(tparams)
       }
-      val kenv0 = getKindEnvFromTypeParamsDefaultStar(tparams1)
-      val kenvVal = kenv0 + (tparams1.tparams.last.sym -> Kind.Eff)
+      val kenv0 = getKindEnvFromTypeParamsDefaultStar(tparams2)
+      val kenvVal = kenv0 + (tparams2.tparams.last.sym -> Kind.Eff)
 
       flatMapN(kenvVal) {
         case kenv =>
-          val tparamsVal = traverse(tparams1.tparams)(visitTypeParam(_, kenv))
+          val tparamsVal = traverse(tparams2.tparams)(visitTypeParam(_, kenv))
 
           flatMapN(tparamsVal) {
-            case tparams =>
+            case tparams0 =>
+              val KindedAst.TypeParam(name, ksym, tloc) = tparams0.last
+              val lastTparam = KindedAst.TypeParam(name, ksym.withKind(Kind.Eff), tloc)
```

Comment:
It seems like we're kind of doing the same thing twice here.

If we've established in the kenv above that the last param is Eff, then we shouldn't have to force that in this line.